src/app/services/baw-api/website-status/website-status.service.ts

```diff
@@ -1,48 +1,74 @@
+  first,
+  interval,
+  map,
+  throwError,
+} from "rxjs";
 
 @Injectable()
-export class WebsiteStatusService implements ApiShow<WebsiteStatus> {
-  public constructor(
-    private http: HttpClient,
-    @Inject(API_ROOT) private apiRoot: string
-  ) {}
+export class WebsiteStatusService {
+  public constructor(private api: BawApiService<WebsiteStatus>) {
+    this.updateStatus();
+    this.tick$.subscribe(() => this.updateStatus());
+  }
+
+  public status$: BehaviorSubject<WebsiteStatus> = new BehaviorSubject(null);
+  private tick$ = interval(30_000); // 30 seconds
+
+  // by making this a service level field, we don't have to recreate the headers every time we make a request
+  private readonly requestHeaders = new HttpHeaders({
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    Accept: "application/json",
+  });
 
   // if the server is fully down, we will not receive a response from the API
   // we can therefore not determine the status of the server and we should return "null"
-  public show(): Observable<WebsiteStatus | null> {
-    const endpoint = `${this.apiRoot}/status`;
+  private show(
+    cacheOptions?: ContextOptions
+  ): Observable<WebsiteStatus | null> {
+    const endpoint = "/status";
 
-    return this.http
-      .get<WebsiteStatus>(endpoint, {
-        headers: WebsiteStatusService.requestHeaders,
+    // we use our custom api.httpGet so that this service using our site wide
+    // caching config (and can be overwritten by the caller in a uniform manner)
+    return (
+      this.api.httpGet(
+        endpoint,
+        this.requestHeaders,
+        cacheOptions
+      ) as any
+    ).pipe(
+      map((response: IWebsiteStatus) => new WebsiteStatus(response)),
+      catchError((err: BawApiError | Error) => {
+        const bawError = isBawApiError(err)
+          ? err
+          : new BawApiError(unknownErrorCode, err.message);
+
+        console.error("Error fetching API /status endpoint:", bawError.message);
+
+        return throwError((): BawApiError => bawError);
       })
-      .pipe(
-        map((response) => new WebsiteStatus(response)),
-        catchError((err: BawApiError | Error) => {
-          const bawError = isBawApiError(err)
-            ? err
-            : new BawApiError(unknownErrorCode, err.message);
-
-          console.error("Error fetching API /status endpoint:", bawError.message);
-
-          return throwError((): BawApiError => bawError);
-        })
-      );
+    );
   }
 
-  // by making this a static field, we don't have to recreate the headers every time we make a request
-  private static readonly requestHeaders = new HttpHeaders({
-    // eslint-disable-next-line @typescript-eslint/naming-convention
-    Accept: "application/json",
-  });
+  private updateStatus(): void {
+    this.show()
+      .pipe(first())
+      .subscribe((status) => {
+        if (this.status$.value !== status) {
+          this.status$.next(status);
+        }
+      });
```

Comment:
Good idea

I've modified it a bit (we use `switchMap` instead of `map` because `this.show` returns an observable)

and I need to set an initial value to `status$`, otherwise we won't know the status until the interval fires its first event (after 30 seconds).

![image](https://github.com/QutEcoacoustics/workbench-client/assets/33742269/e0974daf-52e1-4072-a1db-f6ab20bf8999)

Reply:
For initial trigger you could just do $tick.next?

Also is the wrapper lambda really needed for the show invocation?

Reply:
> For initial trigger you could just do $tick.next?

it's not a behaviour subject, so we can't directly invoke `.next()` on it

![image](https://github.com/QutEcoacoustics/workbench-client/assets/33742269/aeeb0aaf-1f2f-48a0-8c06-e938d53099e0)

We could if we had access to the subscriber [https://rxjs.dev/guide/observable](https://rxjs.dev/guide/observable), but we won't have access to it without some hacky and sub-par code.

I beleive this is the most efficiant and concise way.

> Also is the wrapper lambda really needed for the show invocation?

Kind of:

By using `switchMap(() => this.show())` I'm actually invoking a callback that has the signiture `(interval_time_elapsed: number) => Observable<any>`.

`this.show` used to take `cachingOptions` as it's first parameter meaning it had the signiture `(cache_options: BawCacheOptions) => Observable<WebsiteStatus>` meaning that if I implicitly called it like `switchMap(this.show)`, the `cache_options` would be set to the amount of time elapsed in the interval eg. the number 5 (not good).

However, because I want us to use the `status$` observable, and not the `show()` method, I think it's okay to remove caching options from `this.show()` since `this.show` is a private method, and we never use it and our caching options are already defined by our abstract api (these caching options were to override our defaults).

Reply:
> For initial trigger you could just do $tick.next?

Correction:

By doing the initial call by assigning `this.status$` to `this.show()`, it actually destroyed our binding to `this.tick$`.

I did some looking around, it looks like we can pipe the interval through `startsWith(0)`. This means that when the interval is initially created, it emits an event.

No need for explicit initial event invokation!

---

![image](https://github.com/QutEcoacoustics/workbench-client/assets/33742269/0104befc-0b15-40fd-a6ef-e763de8b08b9)
_New code_